README.md
## Tags and respective `Dockerfile` links

- [`2.6.1`](https://github.com/jdeathe/centos-ssh/releases/tag/2.6.1),`centos-7` [(centos-7/Dockerfile)](https://github.com/jdeathe/centos-ssh/blob/centos-7/Dockerfile)
- [`1.11.1`](https://github.com/jdeathe/centos-ssh/releases/tag/1.11.1),`centos-6` [(centos-6/Dockerfile)](https://github.com/jdeathe/centos-ssh/blob/centos-6/Dockerfile)

## Overview

Included in the build are the [EPEL](http://fedoraproject.org/wiki/EPEL), [IUS](https://ius.io) and [SCL](https://www.softwarecollections.org/) repositories. Installed packages include [inotify-tools](https://github.com/rvoicilas/inotify-tools/wiki), [OpenSSH](http://www.openssh.com/portable.html) secure shell, [Sudo](http://www.courtesan.com/sudo/), [vim-minimal](http://www.vim.org/), python-setuptools, [supervisor](http://supervisord.org/) and [supervisor-stdout](https://github.com/coderanger/supervisor-stdout).

[Supervisor](http://supervisord.org/) is used to start the `sshd` daemon when a docker container based on this image is run.

SSH access is by public key authentication and, by default, the [Vagrant](http://www.vagrantup.com/) [insecure private key](https://github.com/mitchellh/vagrant/blob/master/keys/vagrant) is required.

### Image variants

- [OpenSSH 7.4 / Supervisor 4.0 / EPEL/IUS/SCL Repositories - CentOS-7](https://github.com/jdeathe/centos-ssh/tree/centos-7)
- [OpenSSH 5.3 / Supervisor 3.4 / EPEL/IUS/SCL Repositories - CentOS-6](https://github.com/jdeathe/centos-ssh/tree/centos-6)

### SSH alternatives

SSH is not required in order to access a terminal for the running container. The simplest method is to use the docker exec command to run bash (or sh) as follows:

```
$ docker exec -it {{container-name-or-id}} bash
```

For cases where access to docker exec is not possible the preferred method is to use Command Keys and the nsenter command. See [docs/command-keys.md](https://github.com/jdeathe/centos-ssh/blob/centos-7/docs/command-keys.md) for details on how to set this up.

## Quick start

> For production use, it is recommended to select a specific release tag as shown in the examples.

### SSH mode

Run up an SSH container named 'ssh.1' from the docker image 'jdeathe/centos-ssh' on port 2020 of your docker host.

```
$ docker run -d \
  --name ssh.1 \
  -p 2020:22 \
  jdeathe/centos-ssh:2.6.1
```

Check the logs for the password (required for sudo).

```
$ docker logs ssh.1
```

#### Private key setup

Download the [insecure private key](https://github.com/mitchellh/vagrant/blob/master/keys/vagrant).

```
$ curl -LSs \
  https://raw.githubusercontent.com/mitchellh/vagrant/master/keys/vagrant \
  > id_rsa_insecure
```

Set restrictive permissions on the private key.

```
$ chmod 600 id_rsa_insecure
```
#### Connecting

Connect using the `ssh` command line client with the [insecure private key](https://github.com/mitchellh/vagrant/blob/master/keys/vagrant).

```
$ ssh -p 2020 -i id_rsa_insecure \
  app-admin@{{docker-host-ip}}
```

### SFTP mode

Run up an SFTP container named 'sftp.1' from the docker image 'jdeathe/centos-ssh' on port 2021 of your docker host.

```
$ docker run -d \
  --name sftp.1 \
  -p 2021:22 \
  -e SSH_USER_FORCE_SFTP=true \
  jdeathe/centos-ssh:2.6.1
```

#### Connecting

Connect using the `sftp` command line client with the [insecure private key](https://github.com/mitchellh/vagrant/blob/master/keys/vagrant).

```
$ sftp \
  -o Port=2021 \
  -o StrictHostKeyChecking=no \
  -i id_rsa_insecure \
  app-admin@{{docker-host-ip}}
```

## Instructions

### Running

To run the a docker container from this image you can use the standard docker commands. Alternatively, there's a [docker-compose.yml](https://github.com/jdeathe/centos-ssh/blob/centos-7/docker-compose.yml) example.

For production use, it is recommended to select a specific release tag as shown in the examples.

#### Using scmi

For advanced use-cases, there's an embedded installer (Service Container Manager Interface) [scmi](https://github.com/jdeathe/centos-ssh/blob/centos-7/docs/scmi.md).

#### Using make

If you have a checkout of the [source repository](https://github.com/jdeathe/centos-ssh), and have `make` installed the Makefile provides targets to build, install, start, stop etc; run `make help` for instructions.

#### Using environment variables

The following example overrides the default "app-admin" SSH username, (and corresponding home directory path), with "centos" and "/home/centos" respectively via the `SSH_USER` environment variable.

```
$ docker stop ssh.1 && \
  docker rm ssh.1; \
  docker run -d \
  --name ssh.1 \
  -p :22 \
  --env "SSH_USER=centos" \
  jdeathe/centos-ssh:2.6.1
```

To identify the `SSH_USER` user's sudoer password, inspect the container's logs as follows:

```
$ docker logs ssh.1
```

The output of the logs will show the auto-generated password for the user specified by `SSH_USER` on first run.

```
2019-06-20 00:10:35,306 WARN No file matches via include "/etc/supervisord.d/*.ini"
2019-06-20 00:10:35,306 INFO Included extra file "/etc/supervisord.d/00-supervisor_stdout.conf" during parsing
2019-06-20 00:10:35,307 INFO Included extra file "/etc/supervisord.d/10-system-timezone-wrapper.conf" during parsing
2019-06-20 00:10:35,307 INFO Included extra file "/etc/supervisord.d/20-sshd-bootstrap.conf" during parsing
2019-06-20 00:10:35,307 INFO Included extra file "/etc/supervisord.d/50-sshd-wrapper.conf" during parsing
2019-06-20 00:10:35,307 INFO Set uid to user 0 succeeded
2019-06-20 00:10:35,310 INFO supervisord started with pid 1
2019-06-20 00:10:36,315 INFO spawned: 'system-timezone-wrapper' with pid 9
2019-06-20 00:10:36,318 INFO spawned: 'sshd-bootstrap' with pid 10
2019-06-20 00:10:36,320 INFO spawned: 'sshd-wrapper' with pid 11
INFO: sshd-wrapper waiting on sshd-bootstrap
2019-06-20 00:10:36,328 INFO success: system-timezone-wrapper entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
2019-06-20 00:10:36,328 INFO success: sshd-bootstrap entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)

================================================================================
System Time Zone Details
--------------------------------------------------------------------------------
timezone : UTC
--------------------------------------------------------------------------------
0.00640178

2019-06-20 00:10:36,346 INFO exited: system-timezone-wrapper (exit status 0; expected)

================================================================================
SSH Details
--------------------------------------------------------------------------------
chroot path : N/A
home : /home/app-admin
id : 500:500
key fingerprints :
dd:3b:b8:2e:85:04:06:e9:ab:ff:a8:0a:c0:04:6e:d6 (insecure key)
password : uIEqLkiacCvxaN45
password authentication : no
rsa private key fingerprint :
N/A
rsa host key fingerprint :
7d:6f:d2:e8:7e:84:dd:ff:98:05:5e:6f:35:66:51:53
shell : /bin/bash
sudo : ALL=(ALL) ALL
user : app-admin
--------------------------------------------------------------------------------
0.516901

INFO: sshd-wrapper starting sshd
2019-06-20 00:10:36,852 INFO exited: sshd-bootstrap (exit status 0; expected)
Server listening on 0.0.0.0 port 22.
Server listening on :: port 22.
2019-06-20 00:10:41,872 INFO success: sshd-wrapper entered RUNNING state, process has stayed up for > than 5 seconds (startsecs)
```

#### Environment variables

There are several environmental variables defined at runtime these allow the operator to customise the running container.

##### ENABLE_SSHD_BOOTSTRAP & ENABLE_SSHD_WRAPPER

It may be desirable to prevent the startup of the sshd-bootstrap script and/or sshd daemon. For example, when using an image built from this Dockerfile as the source for another Dockerfile you could disable both sshd-booststrap and sshd from startup by setting `ENABLE_SSHD_BOOTSTRAP` and `ENABLE_SSHD_WRAPPER` to `false`. The benefit of this is to reduce the number of running processes in the final container.

```
...
  --env "ENABLE_SSHD_BOOTSTRAP=false" \
  --env "ENABLE_SSHD_WRAPPER=false" \
...
```

##### ENABLE_SUPERVISOR_STDOUT

This image has `supervisor_stdout` installed which can be used to allow a process controlled by supervisord to send output to both a log file and stdout. It is recommended to simply output to stdout in order to reduce the number of running processes to a minimum. Setting `ENABLE_SUPERVISOR_STDOUT` to "false" will prevent the startup of `supervisor_stdout`. Where an image requires this feature for its logging output `ENABLE_SUPERVISOR_STDOUT` should be set to "true".

##### SSH_AUTHORIZED_KEYS

As detailed below the public key added for the SSH user is insecure by default. This is intentional and allows for access using a known private key. Using `SSH_AUTHORIZED_KEYS` you can replace the insecure public key with another one (or several). Further details on how to create your own private + public key pair are provided below. If adding more than one key it is recommended to either base64 encode the value or use a container file path in combination with a bind mounted file or Docker Swarm config etc.

```
...
--env "SSH_AUTHORIZED_KEYS=
ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEA6NF8iallvQVp22WDkTkyrtvp9eWW6A8YVr+kz4TjGYe7gHzIw+niNltGEFHzD8+v1I2YJ6oXevct1YeS0o9HZyN1Q9qgCgzUFtdOKLv6IedplqoPkcmF0aYet2PkEDo3MlTBckFXPITAMzF8dJSIFo9D8HfdOV0IAdx4O7PtixWKn5y2hMNG0zQPyUecp4pzC6kivAIhyfHilFR61RGL+GPXQ2MWZWFYbAGjyiYJnAmCP3NOTd0jMZEnDkbUvxhMmBYSdETk1rRgm+R4LOzFUGaHqHDLKLX+FIPKcF96hrucXzcWyLbIbEgE98OHlnVYCzRdK8jlqm8tehUc9c9WhQ== vagrant insecure public key
ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAqmLedI2mEJimvIm1OzT1EYJCMwegL/jfsXARLnYkZvJlEHFYDmRgS+YQ+MA9PKHyriCPmVNs/6yVc2lopwPWioXt0+ulH/H43PgB6/4fkP0duauHsRtpp7z7dhqgZOXqdLUn/Ybp0rz0+yKUOBb9ggjE5n7hYyDGtZR9Y11pJ4TuRHmL6wv5mnj9WRzkUlJNYkr6X5b6yAxtQmX+2f33u2qGdAwADddE/uZ4vKnC0jFsv5FdvnwRf2diF/9AagDb7xhZ9U3hPOyLj31H/OUce4xBpGXRfkUYkeW8Qx+zEbEBVlGxDroIMZmHJIknBDAzVfft+lsg1Z06NCYOJ+hSew== another public key
"  \
...
```

> **Note:** The `base64` command on Mac OSX will encode a file without line breaks by default but if using the command on Linux you need to include use the `-w` option to prevent wrapping lines at 80 characters. i.e. `base64 -w 0 -i {{key-path}}`.

```
...
  --env "SSH_AUTHORIZED_KEYS=$(
    cat ${HOME}/.ssh/id_rsa.pub ${HOME}/.ssh/another_id_rsa.pub | base64 -i -
  )" \
...
```

Using `SSH_AUTHORIZED_KEYS` with a container file path allows for the authorized_keys to be populated from the file path.

```
...
  --env "SSH_AUTHORIZED_KEYS=/var/run/config/authorized_keys"
...
```

##### SSH_CHROOT_DIRECTORY

This option is only applicable when `SSH_USER_FORCE_SFTP` is set to `true`. When using the SFTP option the user is jailed into the ChrootDirectory. The value can contain the placeholders `%h` and `%u` which will be replaced with the values of `SSH_USER_HOME` and `SSH_USER` respectively. The default value of `%h` is the best choice in most cases but the user requires a sub-directory in their HOME directory which they have write access to. If no volume is mounted into the path of the SSH user's HOME directory then a directory named `_data` is created automatically. If you need the user to be able to write to their HOME directory then use an alternative value such as `/chroot/%u` so that the user's HOME path, (relative to the ChrootDirectory), becomes `/chroot/app-admin/home/app-admin` by default.

```
...
  --env "SSH_CHROOT_DIRECTORY=%h" \
...
```

##### SSH_INHERIT_ENVIRONMENT

The SSH user's environment is reset by default meaning that the Docker environmental variables are not available. Use `SSH_INHERIT_ENVIRONMENT` to allow the Docker environment variables to be passed to the SSH user's environment. Note that some values are removed to prevent issues; such as SSH_USER_PASSWORD, HOME, HOSTNAME, PATH, TERM etc.

```
...
  --env "SSH_INHERIT_ENVIRONMENT=true" \
...
```

##### SSH_PASSWORD_AUTHENTICATION

The SSH password authentication is disabled by default; allowing access by public/private key based authentication only. This is the recommended configuration however it may be necessary to allow password based access if you have client's that are unable to use key based authentication. Use `SSH_PASSWORD_AUTHENTICATION` to enable password authentication.

```
...
  --env "SSH_PASSWORD_AUTHENTICATION=true" \
...
```

##### SSH_SUDO

On first run the SSH user is created with a the sudo rule `ALL=(ALL)  ALL` which allows the user to run all commands but a password is required. If you want to limit the access to specific commands or allow sudo without a password prompt `SSH_SUDO` can be used.

```
...
  --env "SSH_SUDO=ALL=(ALL) NOPASSWD:ALL" \
...
```

##### SSH_USER

On first run the SSH user is created with the default username of "app-admin". If you require an alternative username `SSH_USER` can be used when running the container.

```
...
  --env "SSH_USER=centos" \
...
```

##### SSH_USER_FORCE_SFTP

To force the use of the internal-sftp command set `SSH_USER_FORCE_SFTP` to `true`. This will prevent shell access, remove the ability to use `sudo` and restrict the user to the ChrootDirectory set using `SSH_SHROOT_DIRECTORY`. Using SFTP in combination with --volumes-from another running container can be used to allow write access to an applications data volume - for example using the `SSH_USER_HOME` value to `/var/www/` could be used to allow access to the data volume of an Apache container.

```
...
  --env "SSH_USER_FORCE_SFTP=false" \
...
```

##### SSH_USER_HOME

On first run the SSH user is created with the default HOME directory of `/home/%u` where `%u` is replaced with the value of `SSH_USER`. If you require an alternative HOME directory `SSH_USER_HOME` can be used when running the container.

```
...
  --env "SSH_USER_HOME=/home/centos" \
...
```

##### SSH_USER_PASSWORD

On first run the SSH user is created with a generated password. If you require a specific password `SSH_USER_PASSWORD` can be used when running the container. If set to an empty string then a password is auto-generated and, if `SSH_SUDO` is not set to allow no password for all commands, will be displayed in the docker logs.

```
...
  --env "SSH_USER_PASSWORD=Passw0rd!" \
...
```

If set to a valid container file path the value will be read from the file - this allows for setting the value securely when combined with an orchestration feature such as Docker Swarm secrets.

```
...
  --env "SSH_USER_PASSWORD=/var/run/secrets/ssh_user_password" \
...
```

##### SSH_USER_PASSWORD_HASHED


CHANGELOG.md
# Change Log

## 2 - centos-7

Summary of release changes.

### 2.6.1 - 2019-09-21

- Deprecate Makefile target `logs-delayed`; replaced with `logsdef`.
- Updates `openssh` package to 7.4p1-21.el7.
- Updates `openssl` package to 1.0.2k-19.el7.
- Updates `sudo` package to 1.8.23-4.el7.
- Updates `yum-plugin-versionlock` package to 1.1.31-52.el7.
- Updates supervisord to 4.0.4.
- Updates `test/health_status` helper script with for consistency.
- Updates Makefile target `logs` to accept `[OPTIONS]` (e.g `make -- logs -ft`).
- Updates `healthcheck` script; state file existence confirms bootstrap completion.
- Updates `system-timezone-wrapper` to improve timer accuracy.
- Updates scripts to explicitly check for a file when handling lock/state files.
- Updates method used for returning current script.
- Updates info/error output for consistency.
- Updates healthcheck failure messages to remove EOL character that is rendered in status response.
- Updates wrapper script; only emit "waiting on" info message if bootstrap hasn't completed.
- Updates CHANGELOG.md to simplify maintenance.
- Updates README.md to simplify contents and improve readability.
- Updates README-short.txt to apply to all image variants.
- Updates Dockerfile `org.deathe.description` metadata LABEL for consistency.
- Updates ordering of Tags and respective Dockerfile links in README.md for readability.
- Adds improved test workflow; added `test-setup` target to Makefile.
- Adds Makefile target `logsdef` to handle deferred logs output within a target chain.
- Adds exec proxy function to `sshd-wrapper` used to pass through nice.
- Adds double quotes around value containing spaces.
- Adds `/docs` directory for supplementary documentation and simplify README.
- Fixes validation failure of 0 second --timeout value in `test/health_status`.
- Removes `ENABLE_SSHD_BOOTSTRAP` from docker-compose example configuration.
- Removes `ENABLE_SSHD_WRAPPER` from docker-compose example configuration.

### 2.6.0 - 2019-06-20

- Deprecates `SSH_AUTOSTART_SSHD`, replaced with `ENABLE_SSHD_WRAPPER`.
- Deprecates `SSH_AUTOSTART_SSHD_BOOTSTRAP`, replaced with `ENABLE_SSHD_BOOTSTRAP`.
- Deprecates `SSH_AUTOSTART_SUPERVISOR_STDOUT`, replaced with `ENABLE_SUPERVISOR_STDOUT`.
- Deprecates `SSH_TIMEZONE`, replaced with `SYSTEM_TIMEZONE`.
- Updates source tag to CentOS 7.6.1810.
- Updates supervisord to 4.0.3.
- Updates default value of `ENABLE_SUPERVISOR_STDOUT` to false.
- Updates `sshd-bootstrap` and `sshd-wrapper` configuration to send error log output to stderr.
- Updates order of values in SSH/SFTP Details log output.
- Updates bootstrap timer to use UTC date timestamps.
- Updates bootstrap supervisord configuration file/priority to `20-sshd-bootstrap.conf`/`20`.
- Updates wrapper supervisord configuration file/priority to `50-sshd-wrapper.conf`/`50`.
- Adds reference to `python-setuptools` in README; removed in error.
- Adds `inspect`, `reload` and `top` Makefile targets.
- Adds improved lock/state file implementation in bootstrap and wrapper scripts.
- Adds improved `clean` Makefile target; includes exited containers and dangling images.
- Adds improved wait on bootstrap completion in wrapper script.
- Adds `system-timezone` and `system-timezone-wrapper` to handle system time zone setup.
- Adds system time zone validation to healthcheck.
- Fixes port incrementation failures when installing systemd units via `scmi`.
- Fixes etcd port registration failures when installing systemd units via `scmi` with the `--register` option.
- Fixes binary paths in systemd unit files for compatibility with both EL and Ubuntu hosts.
- Fixes use of printf binary instead of builtin in systemd unit files.
- Fixes docker host connection status check in Makefile.
- Fixes make clean error thrown when removing exited containers.
- Removes support for long image tags (i.e. centos-7-2.x.x).
- Removes system time zone setup from `sshd-bootstrap`.
- Removes redundant directory test from `sshd-bootstrap`; state file ensures it's a one-shot process.

### 2.5.1 - 2019-02-28

- Deprecates use of `supervisor_stdout` - the default value of `SSH_AUTOSTART_SUPERVISOR_STDOUT` will be switched to "false" in a future release.
- Updates Dockerfile with combined ADD to reduce layer count in final image.
- Fixes `scmi` installation error when using the `--manager=systemd` option on Ubuntu hosts.
- Fixes issues with failure to install/uninstall systemd units installed with scmi.
- Adds improvement to pull logic in systemd unit install template.
- Adds `docker-compose.yml` to `.dockerignore` to reduce size of build context.
- Adds docker-compose configuration example.
- Adds `SSH_AUTOSTART_SUPERVISOR_STDOUT` to control startup of `supervisor_stdout`.
- Adds drop-in configuration for `supervisor_stdout` in `/etc/supervisord.d/00-supervisor_stdout.conf`.
- Adds improved `healtchcheck`, `sshd-bootstrap` and `sshd-wrapper` scripts.
- Adds validation of `SSH_INHERIT_ENVIRONMENT` values.
- Removes reference to `python-setuptools` from README as it's no longer installed.
- Removes requirement of `supervisor_stdout` for output of supervisord logs to stdout.
- Removes unnecessary configuration file `/etc/sshd-bootstrap.conf`.
- Removes unnecessary environment file `/etc/sshd-bootstrap.env`.

### 2.5.0 - 2019-01-28

- Updates `openssl` package to 1.0.2k-16.el7.
- Updates `sudo` package to 1.8.23-3.el7.
- Updates `yum-plugin-versionlock` package to 1.1.31-50.el7.
- Updates supervisor to 3.3.5.
- Updates validation for `SSH_USER_ID` to allow values in the system ID range.
- Updates and restructures Dockerfile to reduce number of layers in image.
- Updates container naming conventions for `scmi` making the node element optional.
- Updates container naming conventions and readability of `Makefile`.
- Updates `docker logs` output example in README document.
- Updates README instructions following review.
- Updates default HEALTHCHECK interval to 1 second from 0.5.

run.sh
#!/usr/bin/env bash

# Change working directory
DIR_PATH="$( if [[ $( echo "${0%/*}" ) != $( echo "${0}" ) ]]; then cd "$( echo "${0%/*}" )"; fi; pwd )"
if [[ ${DIR_PATH} == */* ]] && [[ ${DIR_PATH} != $( pwd ) ]]; then
	cd ${DIR_PATH}
fi

source run.conf
source docker-helpers.sh

# Configuration volume
if [[ ${VOLUME_CONFIG_ENABLED} == true ]] && ! have_docker_container_name ${VOLUME_CONFIG_NAME}; then

	echo "Creating configuration volume."
	if [[ ${VOLUME_CONFIG_NAMED} == true ]]; then
		DOCKER_VOLUME_MAPPING=${VOLUME_CONFIG_NAME}:/etc/services-config
	else
		DOCKER_VOLUME_MAPPING=/etc/services-config
	fi

	(
	set -x
	docker run \
		--name ${VOLUME_CONFIG_NAME} \
		-v ${DOCKER_VOLUME_MAPPING} \
		${DOCKER_IMAGE_REPOSITORY_NAME} \
		/bin/true;
	)

	# Named data volumes require files to be copied into place.
	if [[ ${VOLUME_CONFIG_NAMED} == true ]]; then
		echo "Populating configuration volume."
		(
		set -x
		docker cp \
			./etc/services-config/. \
			${DOCKER_VOLUME_MAPPING};
		)
	fi
fi

# Application container
remove_docker_container_name ${DOCKER_NAME}

if [[ ${#} -eq 0 ]]; then
	echo "Running container ${DOCKER_NAME} as a background/daemon process."
	DOCKER_OPERATOR_OPTIONS="-d"
else
	# This is useful for running commands like 'export' or 'env' to check the
	# environment variables set by the --link docker option.
	#
	# If you need to pipe to another command, quote the commands. e.g:
	#   ./run.sh "env | grep MYSQL | sort"
	printf "Running container %s with CMD [/bin/bash -c '%s']\n" "${DOCKER_NAME}" "${*}"
	DOCKER_OPERATOR_OPTIONS="-it --entrypoint /bin/bash --env TERM=${TERM:-xterm}"
fi

if [[ ${VOLUME_CONFIG_ENABLED} == true ]] && have_docker_container_name ${VOLUME_CONFIG_NAME}; then
	DOCKER_VOLUMES_FROM="--volumes-from ${VOLUME_CONFIG_NAME}"
fi

# In a sub-shell set xtrace - prints the docker command to screen for reference
(
set -xe
docker run \
	${DOCKER_OPERATOR_OPTIONS} \
	--name ${DOCKER_NAME} \
	-p ${DOCKER_PORT_MAP_TCP_22:-}:22 \
	--env "SSH_AUTHORIZED_KEYS=${SSH_AUTHORIZED_KEYS}" \
	--env "SSH_INHERIT_ENVIRONMENT=${SSH_INHERIT_ENVIRONMENT}" \
	--env "SSH_CHROOT_DIRECTORY=${SSH_CHROOT_DIRECTORY}" \
	--env "SSH_SUDO=${SSH_SUDO}" \
	--env "SSH_USER=${SSH_USER}" \
	--env "SSH_USER_PASSWORD_HASHED=${SSH_USER_PASSWORD_HASHED}" \
	--env "SSH_USER_PASSWORD=${SSH_USER_PASSWORD}" \
	--env "SSH_USER_FORCE_SFTP=${SSH_USER_FORCE_SFTP}" \
	--env "SSH_USER_HOME=${SSH_USER_HOME}" \
	--env "SSH_USER_SHELL=${SSH_USER_SHELL}" \
	--env "SSH_USER_ID=${SSH_USER_ID}" \
	${DOCKER_VOLUMES_FROM:-} \
	${DOCKER_IMAGE_REPOSITORY_NAME}${@:+ -c }"${@}"
)

# Forced SFTP
# To connect: sftp -P 2021 -i ~/.ssh/id_rsa_insecure app-sftp@docker-host
# if [[ -n ${DOCKER_PORT_MAP_TCP_22} ]]; then
# 	(( DOCKER_PORT_MAP_TCP_22 ++ ))
# fi
# 
# DOCKER_NAME="${DOCKER_NAME//${SERVICE_UNIT_NAME}/sftp}"
# DOCKER_NAME="${DOCKER_NAME//.${SERVICE_UNIT_LOCAL_ID}./.$(( ${SERVICE_UNIT_LOCAL_ID} + 1 )).}"
# (( SERVICE_UNIT_LOCAL_ID ++ ))
# remove_docker_container_name ${DOCKER_NAME}
# 
# (
# set -xe
# docker run \
# 	${DOCKER_OPERATOR_OPTIONS} \
# 	--name ${DOCKER_NAME} \
# 	-p ${DOCKER_PORT_MAP_TCP_22:-}:22 \
# 	--env "SSH_USER=app-sftp" \
# 	--env "SSH_USER_FORCE_SFTP=true" \
# 	${DOCKER_VOLUMES_FROM:-} \
# 	-v volume-data.${DOCKER_NAME}.ssh-host-keys:/etc/ssh \
# 	-v volume-data.${DOCKER_NAME}.user-home:/home/app-sftp/_data \
# 	${DOCKER_IMAGE_REPOSITORY_NAME}${@:+ -c }"${@}"
# )

# Forced SFTP + apache-php linked volume + persistent host keys
# if [[ -n ${DOCKER_PORT_MAP_TCP_22} ]]; then
# 	(( DOCKER_PORT_MAP_TCP_22 ++ ))
# fi
# 
# DOCKER_NAME="${DOCKER_NAME//${SERVICE_UNIT_NAME}/sftp}"
# DOCKER_NAME="${DOCKER_NAME//.${SERVICE_UNIT_LOCAL_ID}./.$(( ${SERVICE_UNIT_LOCAL_ID} + 1 )).}"
# (( SERVICE_UNIT_LOCAL_ID ++ ))
# remove_docker_container_name ${DOCKER_NAME}
# 
# (
# set -xe
# docker run \
# 	${DOCKER_OPERATOR_OPTIONS} \
# 	--name ${DOCKER_NAME} \
# 	-p ${DOCKER_PORT_MAP_TCP_22:-}:22 \
# 	--env "SSH_CHROOT_DIRECTORY=%h" \
# 	--env "SSH_USER=app-sftp" \
# 	--env "SSH_USER_FORCE_SFTP=true" \
# 	--env "SSH_USER_HOME=/var/www/" \
# 	--env "SSH_USER_ID=501:502" \
# 	${DOCKER_VOLUMES_FROM:-} \
# 	--volumes-from apache-php.app-1.1.1 \
# 	-v volume-data.${DOCKER_NAME}.ssh-host-keys:/etc/ssh \
# 	${DOCKER_IMAGE_REPOSITORY_NAME}${@:+ -c }"${@}"
# )

# Forced SFTP + apache-php linked volume (writeable home directory)
# if [[ -n ${DOCKER_PORT_MAP_TCP_22} ]]; then
# 	(( DOCKER_PORT_MAP_TCP_22 ++ ))
# fi
# 
# DOCKER_NAME="${DOCKER_NAME//${SERVICE_UNIT_NAME}/sftp}"
# DOCKER_NAME="${DOCKER_NAME//.${SERVICE_UNIT_LOCAL_ID}./.$(( ${SERVICE_UNIT_LOCAL_ID} + 1 )).}"
# (( SERVICE_UNIT_LOCAL_ID ++ ))
# remove_docker_container_name ${DOCKER_NAME}
# 
# (
# set -xe
# docker run \
# 	${DOCKER_OPERATOR_OPTIONS} \
# 	--name ${DOCKER_NAME} \
# 	-p ${DOCKER_PORT_MAP_TCP_22:-}:22 \
# 	--env "SSH_CHROOT_DIRECTORY=/var/www" \
# 	--env "SSH_USER=app-sftp" \
# 	--env "SSH_USER_FORCE_SFTP=true" \
# 	--env "SSH_USER_HOME=/app" \
# 	--env "SSH_USER_ID=502:502" \
# 	${DOCKER_VOLUMES_FROM:-} \
# 	--volumes-from apache-php.app-1.1.1 \
# 	-v volume-data.${DOCKER_NAME}.ssh-host-keys:/etc/ssh \
# 	${DOCKER_IMAGE_REPOSITORY_NAME}${@:+ -c }"${@}"
# )

# Use environment variables instead of configuration volume
# SHA-512 hashed password: Passw0rd!
# Salt: salt/pepper.pot.
# if [[ -n ${DOCKER_PORT_MAP_TCP_22} ]]; then
# 	(( DOCKER_PORT_MAP_TCP_22 ++ ))
# fi
# 
# DOCKER_NAME="${DOCKER_NAME//sftp/${SERVICE_UNIT_NAME}}"
# DOCKER_NAME="${DOCKER_NAME//.${SERVICE_UNIT_LOCAL_ID}./.$(( ${SERVICE_UNIT_LOCAL_ID} + 1 )).}"
# (( SERVICE_UNIT_LOCAL_ID ++ ))
# remove_docker_container_name ${DOCKER_NAME}
# 
# (
# set -xe
# docker run \
# 	${DOCKER_OPERATOR_OPTIONS} \
# 	--name ${DOCKER_NAME} \
# 	-p ${DOCKER_PORT_MAP_TCP_22:-}:22 \
# 	--env "SSH_AUTHORIZED_KEYS=
# ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEA6NF8iallvQVp22WDkTkyrtvp9eWW6A8YVr+kz4TjGYe7gHzIw+niNltGEFHzD8+v1I2YJ6oXevct1YeS0o9HZyN1Q9qgCgzUFtdOKLv6IedplqoPkcmF0aYet2PkEDo3MlTBckFXPITAMzF8dJSIFo9D8HfdOV0IAdx4O7PtixWKn5y2hMNG0zQPyUecp4pzC6kivAIhyfHilFR61RGL+GPXQ2MWZWFYbAGjyiYJnAmCP3NOTd0jMZEnDkbUvxhMmBYSdETk1rRgm+R4LOzFUGaHqHDLKLX+FIPKcF96hrucXzcWyLbIbEgE98OHlnVYCzRdK8jlqm8tehUc9c9WhQ== vagrant insecure public key
# ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAqmLedI2mEJimvIm1OzT1EYJCMwegL/jfsXARLnYkZvJlEHFYDmRgS+YQ+MA9PKHyriCPmVNs/6yVc2lopwPWioXt0+ulH/H43PgB6/4fkP0duauHsRtpp7z7dhqgZOXqdLUn/Ybp0rz0+yKUOBb9ggjE5n7hYyDGtZR9Y11pJ4TuRHmL6wv5mnj9WRzkUlJNYkr6X5b6yAxtQmX+2f33u2qGdAwADddE/uZ4vKnC0jFsv5FdvnwRf2diF/9AagDb7xhZ9U3hPOyLj31H/OUce4xBpGXRfkUYkeW8Qx+zEbEBVlGxDroIMZmHJIknBDAzVfft+lsg1Z06NCYOJ+hSew==
# " \
# 	--env "SSH_INHERIT_ENVIRONMENT=true" \
# 	--env "SSH_CHROOT_DIRECTORY=%h" \
# 	--env "SSH_SUDO=ALL=(ALL) ALL" \
# 	--env "SSH_USER=app-1" \
# 	--env "SSH_USER_PASSWORD_HASHED=true" \
# 	--env 'SSH_USER_PASSWORD=$6$salt/pepper.pot.$vXFjBSve4gdT2gmS3p4pXycFSmkN4yT6eE.FmuFTqiSzH1bRFzulKtlYmJIMvP0pfrL4rx6L78ZQ7hjbWNRff1' \
# 	--env "SSH_USER_FORCE_SFTP=false" \
# 	--env "SSH_USER_HOME=/home/app" \
# 	--env "SSH_USER_SHELL=/bin/sh" \
# 	--env "SSH_USER_ID=500:500" \
# 	${DOCKER_VOLUMES_FROM:-} \
# 	-v volume-data.${DOCKER_NAME}.ssh-host-keys:/etc/ssh \
# 	${DOCKER_IMAGE_REPOSITORY_NAME}${@:+ -c }"${@}"
# )

if is_docker_container_name_running ${DOCKER_NAME}; then
	printf -- "\n%s:\n" 'Docker container status'
	show_docker_container_name_status ${DOCKER_NAME}
	printf -- " ${COLOUR_POSITIVE}--->${COLOUR_RESET} %s\n" 'Container running'
elif [[ ${#} -eq 0 ]]; then
	printf -- " ${COLOUR_NEGATIVE}--->${COLOUR_RESET} %s\n" 'ERROR'
	exit 1
fi

# Linked container test
# if [[ ${#} -eq 0 ]]; then

# 	DOCKER_NAME_LINK_HOST=${DOCKER_NAME}.link-host

# 	if [[ -n ${DOCKER_PORT_MAP_TCP_22} ]]; then
# 		(( DOCKER_PORT_MAP_TCP_22 ++ ))
# 	fi

# 	remove_docker_container_name ${DOCKER_NAME_LINK_HOST}

# 	(
# 	set -xe
# 	docker run \
# 		${DOCKER_OPERATOR_OPTIONS} \
# 		--name ${DOCKER_NAME_LINK_HOST} \
# 		-p ${DOCKER_PORT_MAP_TCP_22:-}:22 \
# 		--link ${DOCKER_NAME}:link-guest \
# 		--env "SSH_INHERIT_ENVIRONMENT=true" \
# 		${DOCKER_VOLUMES_FROM:-} \
# 		${DOCKER_IMAGE_REPOSITORY_NAME}${@:+ -c }"${@}"
# 	)

# 	if is_docker_container_name_running ${DOCKER_NAME_LINK_HOST}; then
# 		printf -- "\n%s:\n" 'Docker process status'
# 		show_docker_container_name_status ${DOCKER_NAME_LINK_HOST}
# 		printf -- " ${COLOUR_POSITIVE}--->${COLOUR_RESET} %s\n" 'Container running'
# 	else
# 		printf -- " ${COLOUR_NEGATIVE}--->${COLOUR_RESET} %s\n" 'ERROR'
# 	fi
# fi
